### PR TITLE
Node prediction

### PR DESCRIPTION
**Pull request by ekosman**

### CODE REVIEW

File: src/archs/SimpleGCN.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+class SimpleGCN(nn.Module):
+    def __init__(self, num_node_features, num_classes):
+        super(SimpleGCN, self).__init__()
+        hidden_channels = 16
+        self.conv1 = GATConv(num_node_features, 32)
+        self.conv2 = GATConv(32, 64)
+        self.conv3 = GATConv(64, 128)
+        self.conv4 = GATConv(128, 256)
+        # self.conv3 = GCNConv(hidden_channels, hidden_channels)
+        # self.conv1 = SplineConv(1, 32, dim=1, kernel_size=5)
```
Comment (joeloren) [inline src/archs/SimpleGCN.py]: remove

File: src/archs/aggregators.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+
+AGGREGATORS = {
+    'sum': aggregate_sum,
+    'mean': aggregate_mean,
+    'min': aggregate_min,
+    'max': aggregate_max,
+    'var': aggregate_var,
+    'std': aggregate_std,
+}
```
Comment (joeloren) [inline src/archs/aggregators.py:42]: nit: add a new line

File: src/archs/gat_sage.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import torch
```
Comment (joeloren) [inline src/archs/gat_sage.py:1]: Add a reference to the source code. I believe that this is what you relied on:
https://github.com/rusty1s/pytorch_geometric/blob/master/examples/reddit.py

File: src/archs/gat_sage.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    def forward(self, x, adjs):
+        for i, (edge_index, _, size) in enumerate(adjs):
+            x_target = x[:size[1]]  # Target nodes are always placed first.
+            x = self.convs[i]((x, x_target), edge_index)
+            if i != self.num_layers - 1:
+                x = F.elu(x)
+        return x.log_softmax(dim=-1)
+
+    def inference(self, x_all, subgraph_loader, device):
```
Comment (joeloren) [inline src/archs/gat_sage.py]: Please add some type annotation and/or a docstring, describing the type (and possibly the dimension) of each argument).

File: src/archs/gat_sage.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                    x = F.relu(x)
+                xs.append(x.cpu())
+
+                pbar.update(batch_size)
+
+            x_all = torch.cat(xs, dim=0)
+
+        pbar.close()
+
+        return x_all
```
Comment (joeloren) [inline src/archs/gat_sage.py]: add newline

File: src/utils/evaluate.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        if name in ['mnist']:
+            # Use these custom settings for initialization. Keep the condition as existence in the list in case other
+            # datasets are added in the future
+            self.name = name
+            self.num_tasks = 1
+            self.eval_metric = 'acc'
+        elif name in ['zinc']:
+            self.name = name
+            self.num_tasks = 1
+            self.eval_metric = 'mae'
```
Comment (joeloren) [inline src/utils/evaluate.py]: Simplify (you're assigning the same values in `.name` and `.num_tasks`), and add an `else: raise ValueError("...")` to catch invalid values.

File: src/utils/evaluate.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    def _parse_and_check_input(self, input_dict):
+        if self.eval_metric == 'rocauc' or self.eval_metric == 'ap' or self.eval_metric == 'rmse' or self.eval_metric == 'acc' or self.eval_metric == 'mae':
```
Comment (joeloren) [inline src/utils/evaluate.py]: nit: simplify to `self.eval_metric in ['rocauc', 'ap', 'rmse', 'acc', 'mae']`

File: src/utils/evaluate.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+            # converting to torch.Tensor to numpy on cpu
+            if torch is not None and isinstance(y_true, torch.Tensor):
+                y_true = y_true.detach().cpu().numpy()
+
+            if torch is not None and isinstance(y_pred, torch.Tensor):
+                y_pred = y_pred.detach().cpu().numpy()
+
+            ## check type
+            if not (isinstance(y_true, np.ndarray) and isinstance(y_true, np.ndarray)):
```
Comment (joeloren) [inline src/utils/evaluate.py]: The two conditions are identical. Did you forget to change the second one to `isinstance(y_pred, np.ndarray)`?

File: tst/ogb/main_pyg_with_pruning.py [hunk 9]
```diff
@@ -135,35 +144,76 @@ def register_logging_files(args):
...
-        dataset.transform = transforms.Compose(
-            [augment_edge, lambda data: encode_y_to_arr(data, vocab2idx, args.max_seq_len)])
-        idx2word_mapper = partial(decode_arr_to_seq, idx2vocab=idx2vocab)
+        train_data = list(train_data)
+        test_data = list(test_data)
+
+    elif args.dataset == 'zinc':
+        train_data = ZINC(root='dataset', subset=True, split='train')
+
+        deg = torch.zeros(5, dtype=torch.long)
```
Comment (joeloren) [inline tst/ogb/main_pyg_with_pruning.py]: Where are you using this?

File: tst/ogb/main_pyg_with_pruning.py [hunk 10]
```diff
@@ -135,35 +144,76 @@ def register_logging_files(args):
...
+        validation_data = list(validation_data)
+        test_data = list(test_data)
+
+    elif args.dataset in ['ogbg-molhiv', 'ogbg-molpcba', 'ogbg-ppa', 'ogbg-code2']:
+        dataset = PygGraphPropPredDataset(name=args.dataset, transform=transform)
 
-    return dataset, split_idx, cls_criterion, idx2word_mapper
+        if args.dataset == 'obgb-code2':
+            seq_len_list = np.array([len(seq) for seq in dataset.data.y])
+            print('Target seqence less or equal to {} is {}%.'.format(args.max_seq_len,
```
Comment (joeloren) [inline tst/ogb/main_pyg_with_pruning.py]: nit: switch to using f-strings (`.format()` is Python 2.7 syntax). Also fix typo in "sequence".

File: tst/ogb/main_with_pruning_node_prediction.py [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+
+def main():
+    args = get_args()
+    dataset = get_dataset(args.dataset)
+    data = dataset.data
+    tb_writer = SummaryWriter()
+    tb_writer.iteration = 0
+
+    device = torch.device('cpu')
```
Comment (joeloren) [inline tst/ogb/main_with_pruning_node_prediction.py]: Why did you disable gpu-support?

File: tst/ogb/main_with_pruning_node_prediction.py [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+def get_model(num_features, num_classes, arch):
+    """
+    Retrieves the model corresponding to the given name.
+
+    @param num_features: the dimensionality of the node features
+    @param num_classes: number of target labels
+    @param arch: name of the GNN architecture
+    """
+    if arch == 'sage':
+        model = SAGE(num_features, 256, num_classes)
```
Comment (joeloren) [inline tst/ogb/main_with_pruning_node_prediction.py]: nit: avoid, if you can, using "magic" numbers (256 and 2 in this case). Assign them to variables with meaningful names first, and use those in the function calls. Also, if possible and it makes sense, add them to the ClearML logging.

File: tst/ogb/model_and_data_utils.py [hunk 13]
```diff
@@ -96,4 +124,39 @@ def add_zeros(data: Data) -> Data:
...
+    return data
+
+i=0
+def to_line_graph(data: Data, directed: bool = True) -> Data:
+    """
+    Convert a graph G to its corresponding line-graph L(G)
+    :param data: a torch_gemoetric Data object representing representing a graph
+    """
+    global i
+    print(i)
```
Comment (joeloren) [inline tst/ogb/model_and_data_utils.py]: Are those lines still necessary or are they debug statements that can be safely removed?

File: src/utils/evaluate.py (hunk 8 above)
Comment (ekosman) [reply src/utils/evaluate.py]: I will clarify this with OGB. I don't think it is intended to be this way but they might have a reason

File: src/archs/SimpleGCN.py (hunk 1 above)
Comment (ekosman) [reply src/archs/SimpleGCN.py]: remove the whole file?

File: src/archs/gat_sage.py (hunk 3 above)
Comment (ekosman) [reply src/archs/gat_sage.py:1]: added reference to FastGAT paper. It is implemented following the GAT architecture they use.

File: src/utils/evaluate.py (hunk 8 above)
Comment (ekosman) [reply src/utils/evaluate.py]: They replied. You were right.

**Pull request merged**